workspace/__ardulib__/ST7735/XST7735.cpp

```diff
@@ -0,0 +1,200 @@
+    SPI.endTransaction();
+#endif
+}
+
+void ST7735::setAddrWindow(uint8_t x0, uint8_t y0, uint8_t x1, uint8_t y1) {
+    writeCmd(ST7735_CASET);
+    writeData(0x00);
+    writeData(x0 + _colstart);
+    writeData(0x00);
+    writeData(x1 + _colstart);
+
+    writeCmd(ST7735_RASET);
+    writeData(0x00);
+    writeData(y0 + _rowstart);
+    writeData(0x00);
+    writeData(y1 + _rowstart);
+
+    writeCmd(ST7735_RAMWR);
+}
+
+void ST7735::commandList(const uint8_t* addr) {
+
+    uint8_t numCommands = pgm_read_byte(addr++); // Number of commands to follow
+    uint8_t numArgs;
+    uint16_t ms;
+
+    while (numCommands--) { // For each command...
+        writeCmd(pgm_read_byte(addr++)); // Read, issue command
+        numArgs = pgm_read_byte(addr++); // Number of args to follow
+        ms = numArgs & DELAY; // If hibit set, delay follows args
+        numArgs &= ~DELAY; // Mask out delay bit
+        while (numArgs--) { // For each argument...
+            writeData(pgm_read_byte(addr++)); // Read, issue argument
+        }
+
+        if (ms) {
+            ms = pgm_read_byte(addr++); // Read post-command delay time (ms)
+            if (ms == 255)
+                ms = 500; // If 255, delay for 500 ms
+            delay(ms);
+        }
+    }
+}
+
+void ST7735::commonInit(const uint8_t* cmdList) {
+
+    _colstart = _rowstart = 0; // May be overridden in init func
+
+    pinMode(_rs, OUTPUT);
+    pinMode(_cs, OUTPUT);
+    csport = portOutputRegister(digitalPinToPort(_cs));
+    cspinmask = digitalPinToBitMask(_cs);
+    rsport = portOutputRegister(digitalPinToPort(_rs));
+    rspinmask = digitalPinToBitMask(_rs);
+
+    SPI.begin();
+
+#ifdef SPI_HAS_TRANSACTION
+    spisettings = SPISettings(4000000L, MSBFIRST, SPI_MODE0);
+#else
+#if defined(ARDUINO_ARCH_SAM)
+    SPI.setClockDivider(24); // 4 MHz (half speed)
+#else
+    SPI.setClockDivider(SPI_CLOCK_DIV4); // 4 MHz (half speed)
+#endif
+    SPI.setBitOrder(MSBFIRST);
+    SPI.setDataMode(SPI_MODE0);
+#endif // SPI_HAS_TRANSACTION
+
+    // toggle RST low to reset; CS low so it'll listen to us
+    *csport &= ~cspinmask;
+    if (_rst) {
+        pinMode(_rst, OUTPUT);
+        digitalWrite(_rst, HIGH);
+        delay(500);
+        digitalWrite(_rst, LOW);
+        delay(500);
+        digitalWrite(_rst, HIGH);
+        delay(500);
+    }
+
+    if (cmdList)
+        commandList(cmdList);
+}
+
+void ST7735::init() {
+    switch (_type) {
+        case 0: // G Type
+            commonInit(Gcmd);
+            break;
+        case 1: // B Type
+            commonInit(Bcmd);
+            break;
+        case 2: // RG type
+            commonInit(Rcmd1);
+            commandList(Rcmd2green);
+            _colstart = 2;
+            _rowstart = 1;
+            commandList(Rcmd3);
+            break;
+        case 3: // RR Type
+            commonInit(Rcmd1);
+            commandList(Rcmd2red);
+            commandList(Rcmd3);
+            break;
+        default:
+            break;
+    }
```

Comment:
This code refers to _all_ `XcmdX` globals based on a runtime value `_type`. The compiler has no chance to throw out any of them besides that only 1 of 3 is used in any particular moment. It’s a flash memory waste.

Instead of accepting and holding `_type` in the constructor, make a few initialization methods:

- `initG`
- `initB`
- `initRG`
- `initRR`

Call only the necessary `init` directly in corresponding nodes’ implementation.